src/Surfnet/StepupSelfService/SelfServiceBundle/Service/U2fService.php

```diff
@@ -205,6 +208,69 @@ public function verifyRegistration(VerifyU2fRegistrationCommand $command)
         return RegistrationVerificationResult::success($result['status'], $result['key_handle']);
     }
 
+    public function revokeRegistration(RevokeU2fRegistrationCommand $command)
+    {
+        $this->logger->notice('Revoking U2F registration from U2F verification server');
+
+        $body = [
+            'requester' => ['institution' => $command->institution, 'identity' => $command->identityId],
+            'key_handle' => ['value' => $command->keyHandle],
+        ];
+
+        $response = $this->guzzleClient->post('api/u2f/revoke-registration', ['json' => $body, 'exceptions' => false]);
+        $statusCode = $response->getStatusCode();
+
+        try {
+            $result = $response->json();
+        } catch (\RuntimeException $e) {
+            $this->logger->critical('U2F registration revocation failed; JSON decoding failed.');
+
+            return RegistrationRevocationResult::apiError();
+        }
+
+        $hasErrors = isset($result['errors']) && is_array($result['errors'])
+            && $result['errors'] === array_filter($result['errors'], 'is_string');
+
+        if ($hasErrors && $statusCode >= 400 && $statusCode < 600) {
+            $this->logger->critical(
+                sprintf(
+                    'U2F registration revocation failed; HTTP %d with errors "%s"',
+                    $statusCode,
+                    join(', ', $result['errors'])
+                )
+            );
+
+            return RegistrationRevocationResult::apiError();
```

Comment:
This is unrecoverable, the Middleware has already acted: the actual revocation has been executed.
Failure of the GW api is fault-tolerant: it will not cause issues in the future as a new registration will issue a different keyhandle (at least, to my understanding :wink:).

Hence I think all GW revoke API errors should be properly logged, but not cause errors to be presented to the user.